### DISCUSSION TITLE
[Enhancement] Please get essential things first

### DISCUSSION THREAD

**Discussion by MhAllan** [dotnet/maui#146] [General]
## Summary
Please get essential things ready so we don't need to relay on 3rd parties or wait your fixes too long:

- Take photo, capture video, camera preview, are essential mobile things. what is something small with camera? mobile, right. but until now Xamarin and Xamarin.Essentials both can't open camera in the framework that is supposed to be mobile framework. and the plugin is there for ages, but not included in the framework nor at least enlist it in Nuget under better name than Xam.Plugin.Media.
- Files and Directory Browser
- Multiple file picker
- Image picker and multiple image picker
- Media Player that plays from network streams
- Image manipulation (Wrap Skia/FFImageloading in controls). [here](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/bitmaps/cropping)  for example, why it is not in the framework?
- Propagating gestures (until now a frame with gesture on top will prevent views below it from receiving gestures). also missing gestures like long press.

- You may also revisit the way you prioritise things. I see tasks like "make carousel auto play" is being worked on before fixing many issues that say we actually can't use carousel because it causes crashing, so I don't know the point of autoplaying carousel if the carousel is still too buggy? Another example is [here](https://github.com/dotnet/maui/wiki/Roadmap#net-maui-preview-2-early-2021) we see that "Implementing GlideX" is one of the early features in MAUI, but what are we solving there? are we solving the problem of whoever wants to put 500 image in a non-virtualized layout? We have more than 2300 bugs in XF many of them are about controls being broken. which is more important?

- Make it easy, we would love to help, but we don't have time to build the unbuildable, and track/debug the wrappers of wrappers. Make each class encapsulate it's own logic, no need to too much helpers and inheritance, favour repeating on inheritance and helpers, that will make us contribute better.

**Comment by sps014** [comment]
Also API changes to access Native code with ease for example in Xamarin Form to start custom Android Intent ,dependency injection is only option. Provide much simple API like a popular [Flutter library](https://pub.dev/packages/android_intent) does
```dart
if (platform.isAndroid) {
  AndroidIntent intent = new AndroidIntent(
      action: 'action_view',
      data: 'https://play.google.com/store/apps/details?'
          'id=com.google.android.apps.myapp',
      arguments: {'authAccount': currentUserEmail},
  );
  await intent.launch();
}
```

**Comment by akamud** [comment]
I have to disagree on the “don’t rely on 3rd party libs”. I actually think that having a healthy and active community is essential for a platform/framework’s success. 

I don’t think the team should prioritize something that the community is doing very well. For example, FFImageLoading is an amazing project, why should Forms embed it in the framework, or redo all its work?

Mobile can still be a very constrained platform, I’ll always prefer modularized/composable solutions for optional things instead of baked in. It doesn’t matter if it is made by a 3rd party or not, as long it is well maintained.

**Comment by MhAllan** [comment]
Sure you can't have everything in a framework, and you still need libraries. Why should redo all FFImageLoading? because image processing is the hard one and more essential than you think. You can't call a mobile framework a mobile framework if it can't preload, catch, and circle images. If FFImageLoading stopped supporting your next version of MAUI your app need to reinvent the wheel like what I did with Xamarin Forms when I found that FFImageLoading can't give me good flexible cropping so I created it all from scratch using Skia. Another example, People who used Fody for automatic property changed stayed behind for long time before they can move to .Net standard because Fody was slow to catch up. Third example, remember XLabs? Lastly, the bigger the library the harder to learn and more difficult to find developers want to invest time and effort in them, so we end up with less Xamarin/MAUI enthusiasts.

I don't understand this "Mobile can still be a very constrained platform, I’ll always prefer modularized/composable solutions for optional things instead of baked in" .. why it is baked in if it is done by MS and it is not when done by others?

**Comment by burtonrodman** [comment]
can anyone point me to a library that enables in-app camera preview and capture on Maui for Windows?

i think this is the last major piece missing for my project to adopt Maui.

i see the info about WinUI not implementing CaptureElement yet, so i'm sure that's why it's not ready.
